src/Voron/Data/BTrees/Tree.cs

```diff
@@ -65,53 +67,66 @@ private Tree(LowLevelTransaction llt, Transaction tx, long root, Slice name, boo
                 SetNewPageAllocator(newPageAllocator);
             }
 
-            _recentlyFoundPages = _currentRecentlyFoundPagesPool.Allocate();
+            _recentlyFoundPages = FoundPagesPool.Allocate();
+
+            _state = new TreeMutableState(llt, in header);
+
+            llt.RegisterDisposable(this);
+        }
+
+        private Tree(LowLevelTransaction llt, Transaction tx, Slice name, bool isIndexTree, NewPageAllocator newPageAllocator)
+        {
+            _llt = llt;
+            _tx = tx;
+            IsIndexTree = isIndexTree;
+            Name = name;
 
-            _state = new TreeMutableState(llt)
+            if (newPageAllocator != null)
             {
-                RootPageNumber = root
-            };
+                Debug.Assert(isIndexTree, "If newPageAllocator is set, we must be in a isIndexTree = true");
+                SetNewPageAllocator(newPageAllocator);
+            }
+
+            _recentlyFoundPages = FoundPagesPool.Allocate();
+            _state = new TreeMutableState(llt);
+
+            llt.RegisterDisposable(this);
         }
 
         public Tree(LowLevelTransaction llt, Transaction tx, Slice name, TreeMutableState state)
         {
             _llt = llt;
             _tx = tx;
             Name = name;
-            
-            _recentlyFoundPages = _currentRecentlyFoundPagesPool.Allocate();
 
-            _state = new TreeMutableState(llt);
+            _recentlyFoundPages = FoundPagesPool.Allocate();
+
+
+            _state = new TreeMutableState(llt, state);
             _state = state;
+
+            llt.RegisterDisposable(this);
+        }
+
+        void IDisposable.Dispose()
+        {
+            _recentlyFoundPages.Clear();
+            FoundPagesPool.Free(_recentlyFoundPages);
```

Comment:
In https://github.com/ravendb/ravendb/pull/18159, not a long time ago we removed disposable interface from trees since it can lead to double disposal. Is this safe? I see only usage is internal and by llt, maybe it's better to introduce some custom interface like IPageCachceReleaser just not to be confused with `IDisposable`. 

For me, when I see `IDisposable` that is clear that I have the responsibility of disposal but here is not the case, right?

Reply:
You cannot call `Tree.Dispose` directly, which helps.
But I agree about the intent

Reply:
Yes, but you can write
`using Tree tree = tx.CreateTree("tree");`

without any exceptions.